Query creating new partition will fail with Glue statistics enabled

```
trino:lo_test> create table p (a bigint, b bigint) with (partitioned_by = ARRAY['b']);
CREATE TABLE
trino:lo_test> insert into p values(10, 10);
INSERT: 1 row

Query 20210225_144420_00002_j72cd, FAILED, 1 node
http://localhost:8080/ui/query.html?20210225_144420_00002_j72cd
Splits: 5 total, 5 done (100.00%)
CPU Time: 0.2s total,     0 rows/s,     0B/s, 25% active
Per Node: 0.0 parallelism,     0 rows/s,     0B/s
Parallelism: 0.0
Peak Memory: 0B
5.72 [0 rows, 0B] [0 rows/s, 0B/s]

Query 20210225_144420_00002_j72cd failed: Partition [10] not found (Service: AWSGlue; Status Code: 400; Error Code: EntityNotFoundException; Request ID: e0ee80c2-2960-4918-9a0e-1554350a9c56; Proxy: null)
io.trino.spi.TrinoException: Partition [10] not found (Service: AWSGlue; Status Code: 400; Error Code: EntityNotFoundException; Request ID: e0ee80c2-2960-4918-9a0e-1554350a9c56; Proxy: null)
        at io.trino.plugin.hive.metastore.glue.DefaultGlueColumnStatisticsProvider.updatePartitionStatistics(DefaultGlueColumnStatisticsProvider.java:291)
        at io.trino.plugin.hive.metastore.glue.converter.GlueInputConverter.convertPartition(GlueInputConverter.java:71)
        at io.trino.plugin.hive.metastore.glue.GlueHiveMetastore.lambda$addPartitions$28(GlueHiveMetastore.java:917)
        at io.trino.plugin.hive.metastore.glue.converter.GlueToTrinoConverter.mappedCopy(GlueToTrinoConverter.java:242)
        at io.trino.plugin.hive.metastore.glue.GlueHiveMetastore.lambda$addPartitions$29(GlueHiveMetastore.java:917)
        at io.trino.plugin.hive.metastore.glue.GlueMetastoreApiStats.call(GlueMetastoreApiStats.java:35)
        at io.trino.plugin.hive.metastore.glue.GlueHiveMetastore.addPartitions(GlueHiveMetastore.java:913)
        at io.trino.plugin.hive.metastore.cache.CachingHiveMetastore.addPartitions(CachingHiveMetastore.java:752)
        at io.trino.plugin.hive.metastore.cache.CachingHiveMetastore.addPartitions(CachingHiveMetastore.java:752)
        at io.trino.plugin.hive.HiveMetastoreClosure.addPartitions(HiveMetastoreClosure.java:230)
        at io.trino.plugin.hive.metastore.SemiTransactionalHiveMetastore$PartitionAdder.execute(SemiTransactionalHiveMetastore.java:3328)
        at io.trino.plugin.hive.metastore.SemiTransactionalHiveMetastore$Committer.executeAddPartitionOperations(SemiTransactionalHiveMetastore.java:1989)
        at io.trino.plugin.hive.metastore.SemiTransactionalHiveMetastore.commitShared(SemiTransactionalHiveMetastore.java:1403)
        at io.trino.plugin.hive.metastore.SemiTransactionalHiveMetastore.commit(SemiTransactionalHiveMetastore.java:1158)
        at io.trino.plugin.hive.HiveMetadata.commit(HiveMetadata.java:2935)
        at io.trino.plugin.hive.HiveConnector.commit(HiveConnector.java:212)
        at io.trino.transaction.InMemoryTransactionManager$TransactionMetadata$ConnectorTransactionMetadata.commit(InMemoryTransactionManager.java:594)
        at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
        at com.google.common.util.concurrent.TrustedListenableFutureTask$TrustedFutureInterruptibleTask.runInterruptibly(TrustedListenableFutureTask.java:125)
        at com.google.common.util.concurrent.InterruptibleTask.run(InterruptibleTask.java:69)
        at com.google.common.util.concurrent.TrustedListenableFutureTask.run(TrustedListenableFutureTask.java:78)
        at io.airlift.concurrent.BoundedExecutor.drainQueue(BoundedExecutor.java:80)
        at io.trino.$gen.Trino_unknown____20210225_144315_2.run(Unknown Source)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
        at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: com.amazonaws.services.glue.model.EntityNotFoundException: Partition [10] not found (Service: AWSGlue; Status Code: 400; Error Code: EntityNotFoundException; Request ID: e0ee80c2-2960-4918-9a0e-1554350a9c56; Proxy: null)
        at com.amazonaws.http.AmazonHttpClient$RequestExecutor.handleErrorResponse(AmazonHttpClient.java:1819)
        at com.amazonaws.http.AmazonHttpClient$RequestExecutor.handleServiceErrorResponse(AmazonHttpClient.java:1403)
        at com.amazonaws.http.AmazonHttpClient$RequestExecutor.executeOneRequest(AmazonHttpClient.java:1372)
        at com.amazonaws.http.AmazonHttpClient$RequestExecutor.executeHelper(AmazonHttpClient.java:1145)
        at com.amazonaws.http.AmazonHttpClient$RequestExecutor.doExecute(AmazonHttpClient.java:802)
        at com.amazonaws.http.AmazonHttpClient$RequestExecutor.executeWithTimer(AmazonHttpClient.java:770)
        at com.amazonaws.http.AmazonHttpClient$RequestExecutor.execute(AmazonHttpClient.java:744)
        at com.amazonaws.http.AmazonHttpClient$RequestExecutor.access$500(AmazonHttpClient.java:704)
        at com.amazonaws.http.AmazonHttpClient$RequestExecutionBuilderImpl.execute(AmazonHttpClient.java:686)
        at com.amazonaws.http.AmazonHttpClient.execute(AmazonHttpClient.java:550)
        at com.amazonaws.http.AmazonHttpClient.execute(AmazonHttpClient.java:530)
        at com.amazonaws.services.glue.AWSGlueClient.doInvoke(AWSGlueClient.java:10483)
        at com.amazonaws.services.glue.AWSGlueClient.invoke(AWSGlueClient.java:10450)
        at com.amazonaws.services.glue.AWSGlueClient.invoke(AWSGlueClient.java:10439)
        at com.amazonaws.services.glue.AWSGlueClient.executeUpdateColumnStatisticsForPartition(AWSGlueClient.java:9413)
        at com.amazonaws.services.glue.AWSGlueClient.updateColumnStatisticsForPartition(AWSGlueClient.java:9380)
        at io.trino.plugin.hive.metastore.glue.DefaultGlueColumnStatisticsProvider.lambda$updatePartitionStatistics$9(DefaultGlueColumnStatisticsProvider.java:263)
        at java.base/java.util.concurrent.CompletableFuture$AsyncRun.run(CompletableFuture.java:1736)
        at com.google.common.util.concurrent.DirectExecutor.execute(DirectExecutor.java:30)
        at java.base/java.util.concurrent.CompletableFuture.asyncRunStage(CompletableFuture.java:1750)
        at java.base/java.util.concurrent.CompletableFuture.runAsync(CompletableFuture.java:1959)
        at io.trino.plugin.hive.metastore.glue.DefaultGlueColumnStatisticsProvider.lambda$updatePartitionStatistics$10(DefaultGlueColumnStatisticsProvider.java:263)
        at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
        at java.base/java.util.AbstractList$RandomAccessSpliterator.forEachRemaining(AbstractList.java:720)
        at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
        at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
        at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:913)
        at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
        at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:578)
        at io.trino.plugin.hive.metastore.glue.DefaultGlueColumnStatisticsProvider.updatePartitionStatistics(DefaultGlueColumnStatisticsProvider.java:269)
        ... 25 more
```

--- a/plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/glue/GlueHiveMetastore.java
+++ b/plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/glue/GlueHiveMetastore.java
@@ -130,6 +130,7 @@ import static io.trino.plugin.hive.HiveErrorCode.HIVE_PARTITION_DROPPED_DURING_Q
 import static io.trino.plugin.hive.aws.AwsCurrentRegionHolder.getCurrentRegionFromEC2Metadata;
 import static io.trino.plugin.hive.metastore.MetastoreUtil.makePartitionName;
 import static io.trino.plugin.hive.metastore.MetastoreUtil.verifyCanDropColumn;
+import static io.trino.plugin.hive.metastore.glue.converter.GlueInputConverter.convertPartition;
 import static io.trino.plugin.hive.metastore.glue.converter.GlueToTrinoConverter.mappedCopy;
 import static io.trino.plugin.hive.metastore.thrift.ThriftMetastoreUtil.getHiveBasicStatistics;
 import static io.trino.plugin.hive.metastore.thrift.ThriftMetastoreUtil.updateStatisticsParameters;
@@ -914,7 +915,7 @@ public class GlueHiveMetastore
                 List<Future<BatchCreatePartitionResult>> futures = new ArrayList<>();
 
                 for (List<PartitionWithStatistics> partitionBatch : Lists.partition(partitions, BATCH_CREATE_PARTITION_MAX_PAGE_SIZE)) {
-                    List<PartitionInput> partitionInputs = mappedCopy(partitionBatch, partition -> GlueInputConverter.convertPartition(partition, columnStatisticsProvider));
+                    List<PartitionInput> partitionInputs = mappedCopy(partitionBatch, partition -> GlueInputConverter.convertPartition(partition));
                     futures.add(glueClient.batchCreatePartitionAsync(new BatchCreatePartitionRequest()
                             .withCatalogId(catalogId)
                             .withDatabaseName(databaseName)
@@ -931,6 +932,11 @@ public class GlueHiveMetastore
                         Thread.currentThread().interrupt();
                         throw new TrinoException(HIVE_METASTORE_ERROR, e);
                     }
+                }
+
+                for (PartitionWithStatistics partition : partitions) {
+                    // TODO(https://github.com/trinodb/trino/issues/7033) make updates in batch
+                    columnStatisticsProvider.updatePartitionStatistics(partition.getPartition(), partition.getStatistics().getColumnStatistics());
                 }
 
                 return null;
@@ -987,7 +993,7 @@ public class GlueHiveMetastore
     public void alterPartition(HiveIdentity identity, String databaseName, String tableName, PartitionWithStatistics partition)
     {
         try {
-            PartitionInput newPartition = GlueInputConverter.convertPartition(partition, columnStatisticsProvider);
+            PartitionInput newPartition = convertPartition(partition);
             stats.getAlterPartition().call(() ->
                     glueClient.updatePartition(new UpdatePartitionRequest()
                             .withCatalogId(catalogId)
@@ -995,6 +1001,9 @@ public class GlueHiveMetastore
                             .withTableName(tableName)
                             .withPartitionInput(newPartition)
                             .withPartitionValueList(partition.getPartition().getValues())));
+            columnStatisticsProvider.updatePartitionStatistics(
+                    partition.getPartition(),
+                    partition.getStatistics().getColumnStatistics());
         }
         catch (EntityNotFoundException e) {
             throw new PartitionNotFoundException(new SchemaTableName(databaseName, tableName), partition.getPartition().getValues());

--- a/plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/glue/converter/GlueInputConverter.java
+++ b/plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/glue/converter/GlueInputConverter.java
@@ -28,7 +28,6 @@ import io.trino.plugin.hive.metastore.Partition;
 import io.trino.plugin.hive.metastore.PartitionWithStatistics;
 import io.trino.plugin.hive.metastore.Storage;
 import io.trino.plugin.hive.metastore.Table;
-import io.trino.plugin.hive.metastore.glue.GlueColumnStatisticsProvider;
 
 import java.util.List;
 import java.util.Optional;
@@ -64,11 +63,10 @@ public final class GlueInputConverter
         return input;
     }
 
-    public static PartitionInput convertPartition(PartitionWithStatistics partitionWithStatistics, GlueColumnStatisticsProvider columnStatisticsProvider)
+    public static PartitionInput convertPartition(PartitionWithStatistics partitionWithStatistics)
     {
         PartitionInput input = convertPartition(partitionWithStatistics.getPartition());
         PartitionStatistics statistics = partitionWithStatistics.getStatistics();
-        columnStatisticsProvider.updatePartitionStatistics(partitionWithStatistics.getPartition(), statistics.getColumnStatistics());
         input.setParameters(updateStatisticsParameters(input.getParameters(), statistics.getBasicStatistics()));
         return input;
     }

--- a/plugin/trino-hive/src/test/java/io/trino/plugin/hive/AbstractTestHive.java
+++ b/plugin/trino-hive/src/test/java/io/trino/plugin/hive/AbstractTestHive.java
@@ -878,7 +878,7 @@ public abstract class AbstractTestHive
         return new HiveTransaction(transactionManager, (HiveMetadata) metadataFactory.create());
     }
 
-    interface Transaction
+    protected interface Transaction
             extends AutoCloseable
     {
         ConnectorMetadata getMetadata();
@@ -4776,7 +4776,7 @@ public abstract class AbstractTestHive
         return createTableProperties(storageFormat, ImmutableList.of());
     }
 
-    private static Map<String, Object> createTableProperties(HiveStorageFormat storageFormat, Iterable<String> parititonedBy)
+    protected static Map<String, Object> createTableProperties(HiveStorageFormat storageFormat, Iterable<String> parititonedBy)
     {
         return ImmutableMap.<String, Object>builder()
                 .put(STORAGE_FORMAT_PROPERTY, storageFormat)

--- a/plugin/trino-hive/src/test/java/io/trino/plugin/hive/metastore/glue/TestHiveGlueMetastore.java
+++ b/plugin/trino-hive/src/test/java/io/trino/plugin/hive/metastore/glue/TestHiveGlueMetastore.java
@@ -15,6 +15,7 @@ package io.trino.plugin.hive.metastore.glue;
 
 import com.google.common.collect.ImmutableList;
 import io.airlift.concurrent.BoundedExecutor;
+import io.airlift.slice.Slice;
 import io.trino.plugin.hive.AbstractTestHiveLocal;
 import io.trino.plugin.hive.HiveConfig;
 import io.trino.plugin.hive.HiveMetastoreClosure;
@@ -26,34 +27,51 @@ import io.trino.plugin.hive.metastore.MetastoreConfig;
 import io.trino.plugin.hive.metastore.PartitionWithStatistics;
 import io.trino.plugin.hive.metastore.Table;
 import io.trino.spi.TrinoException;
+import io.trino.spi.block.Block;
 import io.trino.spi.connector.ColumnMetadata;
+import io.trino.spi.connector.ConnectorMetadata;
+import io.trino.spi.connector.ConnectorOutputTableHandle;
+import io.trino.spi.connector.ConnectorPageSink;
+import io.trino.spi.connector.ConnectorSession;
+import io.trino.spi.connector.ConnectorTableMetadata;
 import io.trino.spi.connector.SchemaTableName;
 import io.trino.spi.connector.TableNotFoundException;
 import io.trino.spi.predicate.Domain;
 import io.trino.spi.predicate.Range;
 import io.trino.spi.predicate.TupleDomain;
+import io.trino.spi.statistics.ColumnStatisticMetadata;
+import io.trino.spi.statistics.ComputedStatistics;
+import io.trino.spi.statistics.TableStatisticType;
 import io.trino.spi.type.BigintType;
 import io.trino.spi.type.DateType;
 import io.trino.spi.type.IntegerType;
 import io.trino.spi.type.SmallintType;
 import io.trino.spi.type.TinyintType;
 import io.trino.spi.type.VarcharType;
+import io.trino.testing.MaterializedResult;
 import org.testng.annotations.BeforeClass;
 import org.testng.annotations.Test;
 
 import java.io.File;
 import java.util.ArrayList;
 import java.util.Arrays;
+import java.util.Collection;
 import java.util.List;
 import java.util.Optional;
 import java.util.concurrent.Executor;
 
 import static com.google.common.collect.ImmutableList.toImmutableList;
+import static io.airlift.concurrent.MoreFutures.getFutureValue;
 import static io.airlift.slice.Slices.utf8Slice;
 import static io.trino.plugin.hive.HiveStorageFormat.ORC;
+import static io.trino.plugin.hive.HiveStorageFormat.TEXTFILE;
 import static io.trino.plugin.hive.HiveTestUtils.HDFS_ENVIRONMENT;
 import static io.trino.plugin.hive.metastore.glue.PartitionFilterBuilder.DECIMAL_TYPE;
 import static io.trino.plugin.hive.metastore.glue.PartitionFilterBuilder.decimalOf;
+import static io.trino.spi.statistics.ColumnStatisticType.MAX_VALUE;
+import static io.trino.spi.statistics.ColumnStatisticType.MIN_VALUE;
+import static io.trino.spi.statistics.ColumnStatisticType.NUMBER_OF_DISTINCT_VALUES;
+import static io.trino.spi.statistics.ColumnStatisticType.NUMBER_OF_NON_NULL_VALUES;
 import static io.trino.testing.TestingConnectorSession.SESSION;
 import static java.lang.String.format;
 import static java.util.Locale.ENGLISH;
@@ -674,6 +692,105 @@ public class TestHiveGlueMetastore
                 partitionList,
                 ImmutableList.of(isNullFilter),
                 ImmutableList.of(ImmutableList.of(GlueExpressionUtil.NULL_STRING)));
+    }
+
+    @Test
+    public void testUpdateStatisticsOnCreate()
+    {
+        SchemaTableName tableName = temporaryTable("update_statistics_create");
+        try (Transaction transaction = newTransaction()) {
+            ConnectorSession session = newSession();
+            ConnectorMetadata metadata = transaction.getMetadata();
+
+            List<ColumnMetadata> columns = ImmutableList.of(new ColumnMetadata("a_column", BigintType.BIGINT));
+            ConnectorTableMetadata tableMetadata = new ConnectorTableMetadata(tableName, columns, createTableProperties(TEXTFILE));
+            ConnectorOutputTableHandle createTableHandle = metadata.beginCreateTable(session, tableMetadata, Optional.empty());
+
+            // write data
+            ConnectorPageSink sink = pageSinkProvider.createPageSink(transaction.getTransactionHandle(), session, createTableHandle);
+            MaterializedResult data = MaterializedResult.resultBuilder(session, BigintType.BIGINT)
+                    .row(1L)
+                    .row(2L)
+                    .row(3L)
+                    .row(4L)
+                    .row(5L)
+                    .build();
+            sink.appendPage(data.toPage());
+            Collection<Slice> fragments = getFutureValue(sink.finish());
+
+            // prepare statistics
+            ComputedStatistics statistics = ComputedStatistics.builder(ImmutableList.of(), ImmutableList.of())
+                    .addTableStatistic(TableStatisticType.ROW_COUNT, singleValueBlock(5))
+                    .addColumnStatistic(new ColumnStatisticMetadata("a_column", MIN_VALUE), singleValueBlock(1))
+                    .addColumnStatistic(new ColumnStatisticMetadata("a_column", MAX_VALUE), singleValueBlock(5))
+                    .addColumnStatistic(new ColumnStatisticMetadata("a_column", NUMBER_OF_DISTINCT_VALUES), singleValueBlock(5))
+                    .addColumnStatistic(new ColumnStatisticMetadata("a_column", NUMBER_OF_NON_NULL_VALUES), singleValueBlock(5))
+                    .build();
+
+            // finish CTAS
+            metadata.finishCreateTable(session, createTableHandle, fragments, ImmutableList.of(statistics));
+            transaction.commit();
+        }
+        finally {
+            dropTable(tableName);
+        }
+    }
+
+    @Test
+    public void testUpdatePartitionedStatisticsOnCreate()
+    {
+        SchemaTableName tableName = temporaryTable("update_partitioned_statistics_create");
+        try (Transaction transaction = newTransaction()) {
+            ConnectorSession session = newSession();
+            ConnectorMetadata metadata = transaction.getMetadata();
+
+            List<ColumnMetadata> columns = ImmutableList.of(
+                    new ColumnMetadata("a_column", BigintType.BIGINT),
+                    new ColumnMetadata("part_column", BigintType.BIGINT));
+
+            ConnectorTableMetadata tableMetadata = new ConnectorTableMetadata(tableName, columns, createTableProperties(TEXTFILE, ImmutableList.of("part_column")));
+            ConnectorOutputTableHandle createTableHandle = metadata.beginCreateTable(session, tableMetadata, Optional.empty());
+
+            // write data
+            ConnectorPageSink sink = pageSinkProvider.createPageSink(transaction.getTransactionHandle(), session, createTableHandle);
+            MaterializedResult data = MaterializedResult.resultBuilder(session, BigintType.BIGINT, BigintType.BIGINT)
+                    .row(1L, 1L)
+                    .row(2L, 1L)
+                    .row(3L, 1L)
+                    .row(4L, 2L)
+                    .row(5L, 2L)
+                    .build();
+            sink.appendPage(data.toPage());
+            Collection<Slice> fragments = getFutureValue(sink.finish());
+
+            // prepare statistics
+            ComputedStatistics statistics1 = ComputedStatistics.builder(ImmutableList.of("part_column"), ImmutableList.of(singleValueBlock(1)))
+                    .addTableStatistic(TableStatisticType.ROW_COUNT, singleValueBlock(3))
+                    .addColumnStatistic(new ColumnStatisticMetadata("a_column", MIN_VALUE), singleValueBlock(1))
+                    .addColumnStatistic(new ColumnStatisticMetadata("a_column", MAX_VALUE), singleValueBlock(3))
+                    .addColumnStatistic(new ColumnStatisticMetadata("a_column", NUMBER_OF_DISTINCT_VALUES), singleValueBlock(3))
+                    .addColumnStatistic(new ColumnStatisticMetadata("a_column", NUMBER_OF_NON_NULL_VALUES), singleValueBlock(3))
+                    .build();
+            ComputedStatistics statistics2 = ComputedStatistics.builder(ImmutableList.of("part_column"), ImmutableList.of(singleValueBlock(2)))
+                    .addTableStatistic(TableStatisticType.ROW_COUNT, singleValueBlock(2))
+                    .addColumnStatistic(new ColumnStatisticMetadata("a_column", MIN_VALUE), singleValueBlock(4))
+                    .addColumnStatistic(new ColumnStatisticMetadata("a_column", MAX_VALUE), singleValueBlock(5))
+                    .addColumnStatistic(new ColumnStatisticMetadata("a_column", NUMBER_OF_DISTINCT_VALUES), singleValueBlock(2))
+                    .addColumnStatistic(new ColumnStatisticMetadata("a_column", NUMBER_OF_NON_NULL_VALUES), singleValueBlock(2))
+                    .build();
+
+            // finish CTAS
+            metadata.finishCreateTable(session, createTableHandle, fragments, ImmutableList.of(statistics1, statistics2));
+            transaction.commit();
+        }
+        finally {
+            dropTable(tableName);
+        }
+    }
+
+    private Block singleValueBlock(long value)
+    {
+        return BigintType.BIGINT.createBlockBuilder(null, 1).writeLong(value).build();
     }
 
     private void doGetPartitionsFilterTest(

